packages/@aws-cdk/aws-dynamodb/lib/replica-provider.ts

```diff
@@ -27,10 +27,34 @@ export class ReplicaProvider extends NestedStack {
    */
   public static getOrCreate(scope: Construct, props: ReplicaProviderProps = {}) {
     const stack = Stack.of(scope);
+    this.checkManagedPoliciesLimit(stack);
     const uid = '@aws-cdk/aws-dynamodb.ReplicaProvider';
     return stack.node.tryFindChild(uid) as ReplicaProvider ?? new ReplicaProvider(stack, uid, props);
   }
 
+  public static clearGetOrCreateCalls() {
+    this.getOrCreateCalls.clear();
+  }
+
+  // Map of getOrCreate() calls per stack
+  private static getOrCreateCalls = new Map<string, number>();
+
+  private static checkManagedPoliciesLimit(stack: Stack): void {
+    // The custom resource implementation uses IAM managed policies. There's
+    // a limit of 10 managed policies per role in IAM. Throw if we reach this
+    // limit.
+    const MAX_MANAGED_POLICIES = 10;
+    const calls = this.getOrCreateCalls.get(stack.stackName);
+    if (!calls) {
+      this.getOrCreateCalls.set(stack.stackName, 1);
+    } else {
+      if (calls >= MAX_MANAGED_POLICIES) {
+        throw new Error('The maximum of 10 tables with replication per stack has been reached. Consider splitting your table across multiple stacks.');
+      }
+      this.getOrCreateCalls.set(stack.stackName, calls + 1);
```

Comment:
I feel like this doesn't need a constant, and the 2 calls to `set` are not needed:

```suggestion
    const calls = this.getOrCreateCalls.get(stack.stackName) ?? 0;
    if (calls > 10) {
      throw new Error('You currently cannot have more than 10 global tables in a single stack. Consider splitting your tables across multiple stacks.');
    }
    this.getOrCreateCalls.set(stack.stackName, calls + 1);
```